### PR TITLE
Fix authorization header handling in downloader script.

### PR DESCRIPTION
**Pull request by Syntaxide**
**- What I did**
I've modified the download-frozen-image script, to ensure that the Authorization header is not passed to the blob when downloading. The auth header is only meant for the registry, not the underlying blob storage service.

**- How I did it**

**- How to verify it**
The script currently displays the redirect target. You can confirm that downloads succeed, even when the blob is hosted on s3.
**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by tianon** [comment]
Looks like this needs a DCO sign-off, but I'm also a bit confused.  If we don't need an authorization header for the blob service, couldn't I guess blob values for private, confidential layers and download them without authenticating?  Isn't that why blobs have to be downloaded in the context of a given repo, not simply by their content-addressable-identifier?

Additionally, if all this is doing is removing the extra header, I'm not sure I understand why the diff is so much more than that? :confused:

File: contrib/download-frozen-image-v2.sh [hunk 1]
```diff
@@ -44,6 +44,18 @@ if [ "$(go env GOHOSTOS)" = 'windows' ]; then
...
 fi
 
+fetch_blob() {
+	url=$1
+	token=$2
+	dest=$3
+	echo "Attempting to download blob $url"
+	target=$(curl -v -sS -I -H "Authorization: Bearer $token" "$url" 2>&1 | grep "Location:" | head -n 1 | sed "s/< Location: \([:graph:]*\)/\1/")
+	printf "\n\nFound redirect target: '$target'\n\n"
+	# curl blob (exclude auth token)
```
Comment (thaJeztah) [inline contrib/download-frozen-image-v2.sh]: Ah, looks to be a known issue with cURL; https://stackoverflow.com/questions/37865875/stopping-curl-from-sending-authorization-header-on-302-redirect https://stackoverflow.com/questions/37246261/curl-follow-redirect-without-headers-from-original-request

(Wonder if someone opened an issue with them)

File: contrib/download-frozen-image-v2.sh [hunk 2]
```diff
@@ -158,10 +171,11 @@ while [ $# -gt 0 ]; do
...
 								fi
 								token="$(curl -fsSL "https://auth.docker.io/token?service=registry.docker.io&scope=repository:$image:pull" | jq --raw-output '.token')"
-								curl -fSL --progress \
-									-H "Authorization: Bearer $token" \
-									"https://registry-1.docker.io/v2/$image/blobs/$layerDigest" \
-									-o "$dir/$layerTar"
+								# curl -fSL --progress \
+								# 	-H "Authorization: Bearer $token" \
+								# 	"https://registry-1.docker.io/v2/$image/blobs/$layerDigest" \
+								# 	-o "$dir/$layerTar"
```
Comment (thaJeztah) [inline contrib/download-frozen-image-v2.sh]: Can you remove the commented lines here (and other places)?

**Comment by Syntaxide** [comment]
You provide the header only when talking to registry. Registry does require that you provide the auth token. If and only if the credential is valid, do we issue the redirect. The blob urls are signed and have short ttls, so guessing this url is unlikely.

Curl doesn't have a nice way to extract the redirect, without following the redirect. This extracts the location header, and then makes another request to that url, without the `Authorization` header.

**Comment by tiborvass** [comment]
LGTM Ping @tianon (in case you didn't see because it's collapsed: https://github.com/moby/moby/pull/33707#discussion_r122322764 )

**Comment by thaJeztah** [comment]
Yes @tianon is looking along, and testing

**Comment by andrewhsu** [comment]
LGTM

I've verified this addresses the issue in #33700 download-frozen-image-v2.sh intermittent 400 Bad Request

**Comment by tianon** [comment]
One concern I have with the code as implemented (besides the extra output) is that if the registry changes behavior at some point in the future to no longer redirect and to simply serve the blob data directly, this code is going to break all over again, so I've been playing with different incantations of `curl` to see if we can avoid hard-coding the assumption that it _definitely_ redirects.

**Comment by Syntaxide** [comment]
@tianon How do you feel about having a separate reference implementation of docker pull?

**Comment by tianon** [comment]
I'd love to have a separate reference implementation of `docker pull` -- do you mean in Go?  That's essentially what this script is, just creating a `docker save` tarball without a daemon running.

**Comment by tianon** [comment]
I've had success with the following patch, but I can't reproduce the original failure at all, so I'm not 100% sure it fixes the issue (but it does handle redirects in the required way, and only if there actually _is_ a redirect):
```diff
diff --git a/contrib/download-frozen-image-v2.sh b/contrib/download-frozen-image-v2.sh
index 43bbf7e..8aa2def 100755
--- a/contrib/download-frozen-image-v2.sh
+++ b/contrib/download-frozen-image-v2.sh
@@ -44,6 +44,45 @@ if [ "$(go env GOHOSTOS)" = 'windows' ]; then
 	fi
 fi
 
+registryBase='https://registry-1.docker.io'
+authBase='https://auth.docker.io'
+authService='registry.docker.io'
+#registryBase='https://registry-1-stage.docker.io'
+#authBase='https://auth-stage.docker.io'
+#authService='registry-1-stage.docker.io'
+
+# https://github.com/moby/moby/issues/33700
+fetch_blob() {
+	local token="$1"; shift
+	local image="$1"; shift
+	local digest="$1"; shift
+	local targetFile="$1"; shift
+	local curlArgs=( "$@" )
+
+	local curlHeaders="$(
+		curl -S "${curlArgs[@]}" \
+			-H "Authorization: Bearer $token" \
+			"$registryBase/v2/$image/blobs/$digest" \
+			-o "$targetFile" \
+			-D-
+	)"
+	curlHeaders="$(echo "$curlHeaders" | tr -d '\r')"
+	if [ "$(echo "$curlHeaders" | awk 'NR == 1 { print $2; exit }')" != '200' ]; then
+		rm -f "$targetFile"
+
+		local blobRedirect="$(echo "$curlHeaders" | awk -F ': ' 'tolower($1) == "location" { print $2; exit }')"
+		if [ -z "$blobRedirect" ]; then
+			echo >&2 "error: failed fetching '$image' blob '$digest'"
+			echo "$curlHeaders" | head -1 >&2
+			return 1
+		fi
+
+		curl -fSL "${curlArgs[@]}" \
+			"$blobRedirect" \
+			-o "$targetFile"
+	fi
+}
+
 while [ $# -gt 0 ]; do
 	imageTag="$1"
 	shift
@@ -59,14 +98,14 @@ while [ $# -gt 0 ]; do
 
 	imageFile="${image//\//_}" # "/" can't be in filenames :)
 
-	token="$(curl -fsSL "https://auth.docker.io/token?service=registry.docker.io&scope=repository:$image:pull" | jq --raw-output '.token')"
+	token="$(curl -fsSL "$authBase/token?service=$authService&scope=repository:$image:pull" | jq --raw-output '.token')"
 
 	manifestJson="$(
 		curl -fsSL \
 			-H "Authorization: Bearer $token" \
 			-H 'Accept: application/vnd.docker.distribution.manifest.v2+json' \
 			-H 'Accept: application/vnd.docker.distribution.manifest.v1+json' \
-			"https://registry-1.docker.io/v2/$image/manifests/$digest"
+			"$registryBase/v2/$image/manifests/$digest"
 	)"
 	if [ "${manifestJson:0:1}" != '{' ]; then
 		echo >&2 "error: /v2/$image/manifests/$digest returned something unexpected:"
@@ -87,10 +126,7 @@ while [ $# -gt 0 ]; do
 					imageId="${configDigest#*:}" # strip off "sha256:"
 
 					configFile="$imageId.json"
-					curl -fsSL \
-						-H "Authorization: Bearer $token" \
-						"https://registry-1.docker.io/v2/$image/blobs/$configDigest" \
-						-o "$dir/$configFile"
+					fetch_blob "$token" "$image" "$configDigest" "$dir/$configFile" -s
 
 					layersFs="$(echo "$manifestJson" | jq --raw-output --compact-output '.layers[]')"
 					IFS="$newlineIFS"
@@ -157,11 +193,8 @@ while [ $# -gt 0 ]; do
 									echo "skipping existing ${layerId:0:12}"
 									continue
 								fi
-								token="$(curl -fsSL "https://auth.docker.io/token?service=registry.docker.io&scope=repository:$image:pull" | jq --raw-output '.token')"
-								curl -fSL --progress \
-									-H "Authorization: Bearer $token" \
-									"https://registry-1.docker.io/v2/$image/blobs/$layerDigest" \
-									-o "$dir/$layerTar"
+								token="$(curl -fsSL "$authBase/token?service=$authService&scope=repository:$image:pull" | jq --raw-output '.token')"
+								fetch_blob "$token" "$image" "$layerDigest" "$dir/$layerTar" --progress
 								;;
 
 							*)
@@ -230,8 +263,8 @@ while [ $# -gt 0 ]; do
 					echo "skipping existing ${layerId:0:12}"
 					continue
 				fi
-				token="$(curl -fsSL "https://auth.docker.io/token?service=registry.docker.io&scope=repository:$image:pull" | jq --raw-output '.token')"
-				curl -fSL --progress -H "Authorization: Bearer $token" "https://registry-1.docker.io/v2/$image/blobs/$imageLayer" -o "$dir/$layerId/layer.tar" # -C -
+				token="$(curl -fsSL "$authBase/token?service=$authService&scope=repository:$image:pull" | jq --raw-output '.token')"
+				fetch_blob "$token" "$image" "$imageLayer" "$dir/$layerId/layer.tar" --progress
 			done
 			;;
 
```

I've left in the staging server URLs commented out since they're useful for testing against (given that they should still have the issue).

**Comment by tianon** [comment]
It also preserves the spirit of the original output (one progress bar per layer).  On a redirect, we'll get two progress bars, but IMO that's an acceptable concession.

**Comment by tianon** [comment]
@Syntaxide do you want to continue working on this, or should I carry from here?

**Comment by andrewhsu** [comment]
Let's work on getting this PR adjusted with @tianon 's suggestion so we can port this to the release branches we have.

**Comment by Syntaxide** [comment]
I have confirmed that @tianon 's changes also fix the issue. The test failures appeared to be unrelated to this change, I've rebased to if that fixes the tests.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by andrewhsu** [comment]
LGTM

Stuff works:
```
$ ./contrib/download-frozen-image-v2.sh ./docker-frozen-images \
>   buildpack-deps:jessie@sha256:85b379ec16065e4fe4127eb1c5fb1bcc03c559bd36dbb2e22ff496de55925fa6 \
>   busybox:latest@sha256:32f093055929dbc23dec4d03e09dfe971f5973a9ca5cf059cbfb644c206aa83f \
>   debian:jessie@sha256:72f784399fd2719b4cb4e16ef8e369a39dc67f53d978cd3e2e7bf4e502c7b793 \
>   hello-world:latest@sha256:c5515758d4c5e1e838e9cd307f6c6a0d620b5e07e6f927b07d05f6d12a1ac8d7
Downloading 'library/buildpack-deps:jessie@sha256:85b379ec16065e4fe4127eb1c5fb1bcc03c559bd36dbb2e22ff496de55925fa6' (4 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%

Downloading 'library/busybox:latest@sha256:32f093055929dbc23dec4d03e09dfe971f5973a9ca5cf059cbfb644c206aa83f' (1 layers)...
######################################################################## 100.0%
######################################################################## 100.0%

Downloading 'library/debian:jessie@sha256:72f784399fd2719b4cb4e16ef8e369a39dc67f53d978cd3e2e7bf4e502c7b793' (1 layers)...
skipping existing b6e0d04482ff

Downloading 'library/hello-world:latest@sha256:c5515758d4c5e1e838e9cd307f6c6a0d620b5e07e6f927b07d05f6d12a1ac8d7' (1 layers)...
######################################################################## 100.0%
######################################################################## 100.0%

Download of images into './docker-frozen-images' complete.
Use something like the following to load the result into a Docker daemon:
  tar -cC './docker-frozen-images' . | docker load
```

**Comment by tiborvass** [comment]
LGTM

**Pull request merged**